psqlgraph/edge.py

```diff
@@ -21,63 +23,61 @@ def IDColumn(tablename):
 
 
 class Edge(AbstractConcreteBase, ORMBase):
-
     __src_table__ = None
     __dst_table__ = None
 
     src_id, dst_id, src, dst = None, None, None, None
 
     @declared_attr
     def src_id(cls):
-        if cls.__name__ == 'Edge':
-            return
-        src_table = cls.__src_table__ or NODE_TABLENAME_SCHEME.format(
-            class_name=cls.__src_class__.lower())
-        src_id = IDColumn(src_table)
-        return src_id
+        src_table = cls.__src_table__
+
+        if not src_table and hasattr(cls, "__src_class__"):
+            src_table = NODE_TABLENAME_SCHEME.format(class_name=cls.__src_class__.lower())
+
+        return id_column(src_table, cls.__name__)
 
     @declared_attr
     def dst_id(cls):
-        if cls.__name__ == 'Edge':
-            return
-        dst_table = cls.__dst_table__ or NODE_TABLENAME_SCHEME.format(
-            class_name=cls.__dst_class__.lower())
-        dst_id = IDColumn(dst_table)
-        return dst_id
+
+        dst_table = cls.__dst_table__
+
+        if not dst_table and hasattr(cls, "__dst_class__"):
+            dst_table = NODE_TABLENAME_SCHEME.format(class_name=cls.__dst_class__.lower())
+        return id_column(dst_table, cls.__name__)
 
     @classmethod
     def __declare_last__(cls):
-        if cls == Edge:
-            return
-        assert hasattr(cls, '__src_class__'),\
-            'You must declare __src_class__ for {}'.format(cls)
-        assert hasattr(cls, '__dst_class__'),\
-            'You must declare __dst_class__ for {}'.format(cls)
-        assert hasattr(cls, '__src_dst_assoc__'),\
-            'You must declare __src_dst_assoc__ for {}'.format(cls)
-        assert hasattr(cls, '__dst_src_assoc__'),\
-            'You must declare __dst_src_assoc__ for {}'.format(cls)
+        for scls in Edge.get_subclasses():
```

Comment:
so it used to call it for all subclasses before but now you need to be explicit with it?

Reply:
It used to be called once per sub_class even though it is defined here at them but the behavior changed and its only called once now